AnnService/inc/Core/Common/cuda/KNN.hxx

```diff
@@ -28,245 +28,129 @@
+    if(tptree->leaf_points[leaf_offset+i] >= min_id && tptree->leaf_points[leaf_offset+i] < max_id) {
+      queryId = tptree->leaf_points[leaf_offset + i];
 
-          if(target.dist <= heapMem.top()) {
-            dup=false;
-            for(int dup_id=0; dup_id < KVAL-1; ++dup_id) {
-              if(heapMem.vals[dup_id].idx == target.idx) {
-                dup = true;
-                dup_id=KVAL;
-              }
-            }
-            if(!dup) { // Only consider it if not already in the KNN list
-              max_K.dist = heapMem.vals[0].dist;
-              max_K.idx = heapMem.vals[0].idx;
-              heapMem.insert(target.dist, target.idx);
-            }
-          }
-          else {
-            max_K.dist = target.dist;
-            max_K.idx = target.idx;
-          }
-        }
+      for(int j=0; j<Dim; ++j) {
+        query[j] = ps->getVec(queryId)[j];
       }
-    }
-    // Write KNN to result list in sorted order
-    results[((long long int)query.id+1)*KVAL-1] = max_K.idx;
-    for(int j=KVAL-2; j>=0; j--) {
-      results[(long long int)query.id*KVAL+j] = heapMem.vals[0].idx;
-      heapMem.vals[0].dist=-1;
-      heapMem.heapify();
-    }
-  }
-}
-
-/*****************************************************************************************
- * For a given point, @src_id, looks at all neighbors' neighbors to refine KNN if any nearer
- * neighbors are found.  Recursively continues based on @DEPTH macro value.
- *****************************************************************************************/
-template<typename T, typename SUMTYPE, int Dim, int BLOCK_DIM>
-__device__ void check_neighbors(Point<T,SUMTYPE,Dim>* data, int* results, int KVAL, ThreadHeap<T, SUMTYPE, Dim, BLOCK_DIM>* heapMem, DistPair<SUMTYPE>* max_K, int neighbor, int src_id, Point<T,SUMTYPE,Dim>* query, int metric) {
 
-  DistPair<SUMTYPE> target;
-  bool dup;
-
-  for(long long int j=0; j<KVAL; ++j) { // Check each neighbor of this neighbor
-    target.idx = results[neighbor*KVAL + j];
-    if(target.idx != src_id && target.idx != -1) { // Don't include the source itself
-
-      dup = (target.idx == max_K->idx);
-      for(int dup_id=0; dup_id < KVAL-1 && dup==false; ++dup_id) {
-        dup = (heapMem->vals[dup_id].idx == target.idx);
-      }
-      target.dist = INFTY<SUMTYPE>();
-      if(!dup) {
-        if(metric == 0) {
-          target.dist = query->l2(&data[target.idx]);
-        }
-        else if(metric == 1) {
-          target.dist = query->cosine(&data[target.idx]);
-        }
-      }
-      if(target.dist < max_K->dist) {
-        if(target.dist < heapMem->top()) {
-          max_K->dist = heapMem->vals[0].dist;
-          max_K->idx = heapMem->vals[0].idx;
-          heapMem->insert(target.dist, target.idx);
+      // Load results from previous iterations into shared memory heap
+      // and re-compute distances since they are not stored in result set
```

Comment:
So, we get the result set in order sorted by distance, but not the distance? Is it faster to do the comparison on GPU?

Reply:
Yes, the increased memory requirement of storing the distances is not worth it since it may require more batches.  Re-computing the distances is relatively cheap for GPU.